balancer/rls/internal/keys/builder.go

```diff
@@ -108,16 +114,31 @@ type KeyMap struct {
 
 // RLSKey builds the RLS keys to be used for the given request, identified by
 // the request path and the request headers stored in metadata.
-func (bm BuilderMap) RLSKey(md metadata.MD, path string) KeyMap {
+func (bm BuilderMap) RLSKey(md metadata.MD, host, path string) KeyMap {
+	i := strings.LastIndex(path, "/")
+	service, method := path[:i+1], path[i+1:]
 	b, ok := bm[path]
 	if !ok {
-		i := strings.LastIndex(path, "/")
-		b, ok = bm[path[:i+1]]
+		b, ok = bm[service]
 		if !ok {
 			return KeyMap{}
 		}
 	}
-	return b.keys(md)
+
+	kvMap := b.buildHeaderKeys(md)
+	if b.hostKey != "" {
+		kvMap[b.hostKey] = host
+	}
+	if b.serviceKey != "" {
+		kvMap[b.serviceKey] = service
+	}
+	if b.methodKey != "" {
+		kvMap[b.methodKey] = method
+	}
+	for k, v := range b.constantKeys {
+		kvMap[k] = v
+	}
+	return KeyMap{Map: kvMap, Str: mapToString(kvMap)}
```

Comment:
Do we need to store `Str` in `KeyMap`?  Or can we compute it when we add/retrieve from the map?  (Basically: if we're only consuming `Str` once, we shouldn't store it.)

Reply:
`RLSKey` method will be invoked for every `Pick`, and the returned keys (string form) will be used to lookup the data cache. There is no sharing of the key map between RPCs. So, we can either compute it here or `Pick` will have to compute it right after it calls this method. I don't have much of a preference and I'm leaning towards leaving it as is, to keep `Pick` working the way it currently does. Do you have a preference?

Reply:
For now it seems better to avoid storing it to me, but let's wait for the rest of the changes to decide if it's worth changing.

Reply:
Ack.